catalog/solr/catalog-solr-offline-gazetteer/src/main/java/ddf/catalog/solr/offlinegazetteer/GazetteerQueryOfflineSolr.java

```diff
@@ -0,0 +1,426 @@
+    } catch (SolrServerException | IOException e) {
+      LOGGER.debug("Something went wrong when querying", e);
+      throw new GeoEntryQueryException("Error while querying", e);
+    }
+
+    return Optional.ofNullable(response)
+        .map(QueryResponse::getSuggesterResponse)
+        .map(SuggesterResponse::getSuggestions)
+        .map(suggestionsPerDict -> suggestionsPerDict.get(SUGGEST_PLACE_KEY))
+        .orElse(Collections.emptyList())
+        .stream()
+        .map(suggestion -> new SuggestionImpl(suggestion.getPayload(), suggestion.getTerm()))
+        .collect(Collectors.toList());
+  }
+
+  public static class SuggestionImpl implements Suggestion {
+    private final String id;
+    private final String name;
+
+    public SuggestionImpl(String id, String name) {
+      this.id = id;
+      this.name = name;
+    }
+
+    @Override
+    public String getId() {
+      return id;
+    }
+
+    @Override
+    public String getName() {
+      return name;
+    }
+  }
+
+  @Override
+  public List<NearbyLocation> getNearestCities(String location, int radiusInKm, int maxResults)
+      throws ParseException, GeoEntryQueryException {
+    Geometry geometry;
+    try {
+      geometry = WKT_READER_THREAD_LOCAL.get().read(location);
+    } catch (org.locationtech.jts.io.ParseException e) {
+      throw new GeoEntryQueryException("Could not parse location");
+    }
+    // conver km to rough degree measurement, approximately 111km per degree
+    double distanceInDegrees = radiusInKm / KM_PER_DEGREE;
+    final Geometry originalGeometry = geometry;
+    Geometry bufferedGeo = originalGeometry.buffer(distanceInDegrees, 14);
+    String wkt = WKT_WRITER_THREAD_LOCAL.get().write(bufferedGeo);
+
+    String q =
+        String.format(
+            "location_geo_index:\"Intersects( %s ) AND %s\"",
+            ClientUtils.escapeQueryChars(wkt), CITY_SOLR_QUERY);
+
+    SolrQuery solrQuery = new SolrQuery(q);
+    solrQuery.setRows(Math.min(maxResults, MAX_RESULTS));
+
+    QueryResponse response;
+    try {
+      response = client.query(solrQuery, METHOD.POST);
+    } catch (SolrServerException | IOException e) {
+      LOGGER.debug("Error executing query for nearest cities", e);
+      throw new GeoEntryQueryException("Error executing query", e);
+    }
+
+    return response
+        .getResults()
+        .stream()
+        .map(result -> convert(result, originalGeometry))
+        .collect(Collectors.toList());
+  }
+
+  private NearbyLocation convert(SolrDocument doc, Geometry originalLocation) {
+    String location = getField(doc, "location_geo", String.class);
+    String title =
+        Optional.ofNullable(getField(doc, "title_txt", String.class))
+            .filter(Objects::nonNull)
+            .filter(s -> !"".equals(s))
+            .orElse("NO TITLE");
+
+    String cardinalDirection = "";
+    double distance = 0;
+    try {
+      Geometry geo = WKT_READER_THREAD_LOCAL.get().read(location);
```

Comment:
It _could_ be extracted. But since its only 14 lines of calculation I'd rather opt to leave it in this class rather than introduce an entire other class. If you really want to see this change happen though lets hop on a call and do it together so you can help me see your vision of it 😄